src/Orleans.CodeGenerator/SerializerGenerator.cs

```diff
@@ -412,19 +410,18 @@ private static void AddSerializationMembers(ISerializableTypeDescription type, L
                     codecExpression = IdentifierName(instanceCodec.FieldName);
                 }
 
-                var expectedType = serializerFields.First(f => f is TypeFieldDescription tf && SymbolEqualityComparer.Default.Equals(tf.UnderlyingType, memberType));
-                var writeFieldExpr = ExpressionStatement(
-                        InvocationExpression(
-                            codecExpression.Member("WriteField"),
-                            ArgumentList(
-                                SeparatedList(
-                                    new[]
-                                    {
-                                        Argument(writerParam).WithRefOrOutKeyword(Token(SyntaxKind.RefKeyword)),
-                                        Argument(fieldIdDeltaExpr),
-                                        Argument(expectedType.FieldName.ToIdentifierName()),
-                                        Argument(member.GetGetter(instanceParam))
-                                    }))));
+                var writeFieldArgs = new List<ArgumentSyntax> {
```

Comment:
```suggestion
               // When a static codec is available, we can call it directly and can skip passing the expected type, since it is known to be the static codec's field type:
               //   C#: <staticCodec>.WriteField(ref writer, <fieldIdDelta, <member>)
               // When no static codec is available:
               // C#: <codecField>.WriteField(ref writer, <fieldIdDelta>, <expectedType>, <member>)
                var writeFieldArgs = new List<ArgumentSyntax> {
```